onnxruntime/core/providers/cpu/signal/dft.cc

```diff
@@ -0,0 +1,517 @@
+    0x64, 0xE4, 0x14, 0x94, 0x54, 0xD4, 0x34, 0xB4, 0x74, 0xF4, 0x0C, 0x8C, 0x4C, 0xCC, 0x2C, 0xAC, 0x6C, 0xEC, 0x1C,
+    0x9C, 0x5C, 0xDC, 0x3C, 0xBC, 0x7C, 0xFC, 0x02, 0x82, 0x42, 0xC2, 0x22, 0xA2, 0x62, 0xE2, 0x12, 0x92, 0x52, 0xD2,
+    0x32, 0xB2, 0x72, 0xF2, 0x0A, 0x8A, 0x4A, 0xCA, 0x2A, 0xAA, 0x6A, 0xEA, 0x1A, 0x9A, 0x5A, 0xDA, 0x3A, 0xBA, 0x7A,
+    0xFA, 0x06, 0x86, 0x46, 0xC6, 0x26, 0xA6, 0x66, 0xE6, 0x16, 0x96, 0x56, 0xD6, 0x36, 0xB6, 0x76, 0xF6, 0x0E, 0x8E,
+    0x4E, 0xCE, 0x2E, 0xAE, 0x6E, 0xEE, 0x1E, 0x9E, 0x5E, 0xDE, 0x3E, 0xBE, 0x7E, 0xFE, 0x01, 0x81, 0x41, 0xC1, 0x21,
+    0xA1, 0x61, 0xE1, 0x11, 0x91, 0x51, 0xD1, 0x31, 0xB1, 0x71, 0xF1, 0x09, 0x89, 0x49, 0xC9, 0x29, 0xA9, 0x69, 0xE9,
+    0x19, 0x99, 0x59, 0xD9, 0x39, 0xB9, 0x79, 0xF9, 0x05, 0x85, 0x45, 0xC5, 0x25, 0xA5, 0x65, 0xE5, 0x15, 0x95, 0x55,
+    0xD5, 0x35, 0xB5, 0x75, 0xF5, 0x0D, 0x8D, 0x4D, 0xCD, 0x2D, 0xAD, 0x6D, 0xED, 0x1D, 0x9D, 0x5D, 0xDD, 0x3D, 0xBD,
+    0x7D, 0xFD, 0x03, 0x83, 0x43, 0xC3, 0x23, 0xA3, 0x63, 0xE3, 0x13, 0x93, 0x53, 0xD3, 0x33, 0xB3, 0x73, 0xF3, 0x0B,
+    0x8B, 0x4B, 0xCB, 0x2B, 0xAB, 0x6B, 0xEB, 0x1B, 0x9B, 0x5B, 0xDB, 0x3B, 0xBB, 0x7B, 0xFB, 0x07, 0x87, 0x47, 0xC7,
+    0x27, 0xA7, 0x67, 0xE7, 0x17, 0x97, 0x57, 0xD7, 0x37, 0xB7, 0x77, 0xF7, 0x0F, 0x8F, 0x4F, 0xCF, 0x2F, 0xAF, 0x6F,
+    0xEF, 0x1F, 0x9F, 0x5F, 0xDF, 0x3F, 0xBF, 0x7F, 0xFF};
+
+template <typename T>
+static inline T bit_reverse(T num, unsigned significant_bits) {
+  if (significant_bits > 32) {
+    ORT_THROW("Unsupported bit size.");
+  }
+  uint32_t num_32 = static_cast<uint32_t>(num);
+  uint32_t rev = (BitReverseTable256[num_32 & 0xff] << 24) | (BitReverseTable256[(num_32 >> 8) & 0xff] << 16) |
+                 (BitReverseTable256[(num_32 >> 16) & 0xff] << 8) | (BitReverseTable256[(num_32 >> 24) & 0xff]);
+  return static_cast<T>(((uint64_t)rev) >> (32 - significant_bits));
+}
+
+template <typename T>
+static T compute_angular_velocity(size_t number_of_samples, bool inverse) {
+  // Calculate fundamental angular velocity
+  static const T pi = static_cast<T>(3.14159265);
+  static const T tau = 2 * pi;
+  T inverse_switch = inverse ? 1.f : -1.f;
+  T angular_velocity = inverse_switch * tau / number_of_samples;
+  return angular_velocity;
+}
+
+template <typename T>
+static std::complex<T> compute_exponential(size_t index, const T angular_velocity) {
+  const T angle = static_cast<T>(index) * angular_velocity;
+  return std::complex<T>(cos(angle), sin(angle));
+}
+
+template <typename T, typename U>
+static Status fft_radix2(OpKernelContext* /*ctx*/, const Tensor* X, Tensor* Y, size_t X_offset, size_t X_stride,
+                         size_t Y_offset, size_t Y_stride, int64_t axis, size_t dft_length, const Tensor* window,
+                         bool is_onesided, bool inverse, InlinedVector<std::complex<T>>& V,
+                         InlinedVector<std::complex<T>>& temp_output) {
```

Comment:
@smk2007 there's something wrong with the use of `temp_output` that results in what I think are essentially seg-faults in the WebAssembly build:
```
exiting due to exception: RuntimeError: memory access out of bounds,RuntimeError: memory access out of bounds
    at dlfree (<anonymous>:wasm-function[370078]:0x5d63a08)
    at operator delete(void*) (<anonymous>:wasm-function[371661]:0x5d98073)
    at void std::__2::__libcpp_operator_delete<void*>(void*) (<anonymous>:wasm-function[800]:0xfb6a2)
    at void std::__2::__do_deallocate_handle_size<>(void*, unsigned long) (<anonymous>:wasm-function[798]:0xfb605)
    at std::__2::__libcpp_deallocate(void*, unsigned long, unsigned long) (<anonymous>:wasm-function[795]:0xfb505)
    at invoke_viii (/home/azureuser/src/onnxruntime/build/Linux/Debug/onnxruntime_test_all.js:7113:27)
    at std::__2::allocator<std::__2::complex<float> >::deallocate(std::__2::complex<float>*, unsigned long) (<anonymous>:wasm-function[254957]:0x467f621)
    at std::__2::allocator_traits<std::__2::allocator<std::__2::complex<float> > >::deallocate(std::__2::allocator<std::__2::complex<float> >&, std::__2::complex<float>*, unsigned long) (<anonymous>:wasm-function[254956]:0x467f56f)
    at absl::lts_20211102::inlined_vector_internal::MallocAdapter<std::__2::allocator<std::__2::complex<float> >, false>::Deallocate(std::__2::allocator<std::__2::complex<float> >&, std::__2::complex<float>*, unsigned long) (<anonymous>:wasm-function[254952]:0x467f467)
    at absl::lts_20211102::inlined_vector_internal::Storage<std::__2::complex<float>, 7ul, std::__2::allocator<std::__2::complex<float> > >::DeallocateIfAllocated() (<anonymous>:wasm-function[254931]:0x467e4b6)
    at invoke_vi (/home/azureuser/src/onnxruntime/build/Linux/Debug/onnxruntime_test_all.js:7146:27)
    at absl::lts_20211102::inlined_vector_internal::Storage<std::__2::complex<float>, 7ul, std::__2::allocator<std::__2::complex<float> > >::~Storage() (<anonymous>:wasm-function[254898]:0x46794a1)
    at absl::lts_20211102::InlinedVector<std::__2::complex<float>, 7ul, std::__2::allocator<std::__2::complex<float> > >::~InlinedVector() (<anonymous>:wasm-function[254833]:0x4668238)
    at onnxruntime::common::Status onnxruntime::short_time_fourier_transform<float, float>(onnxruntime::OpKernelContext*, bool, bool) (<anonymous>:wasm-function[254840]:0x466d9d6)
    at onnxruntime::STFT::Compute(onnxruntime::OpKernelContext*) const (<anonymous>:wasm-function[254839]:0x466a637)
```

It seems temp_output is used to avoid re-allocating memory when this is called in a loop?
I'm going to debug this more tomorrow, but if you can spot a bug, please let me know.

Reply:
@smk2007 Can you explain why `temp_output` is needed / helpful in `fft_radix2`?

Reply:
temp output is used in onesided dft case, where the fft gets written to memory but only half of it is copied back to the output buffer. Instead of allocating the same memory again and again, temp output holds onto the dft output, and the memory is reused across invocations.

Reply:
Even if I remove `temp_output` entirely the issue remains with `V`. Since it seems it's only on WebAssembly, and I've stared at the code for some time and don't see anything wrong, I sort of want to blame the compiler, but we're probably doing something that's undefined behavior. I tried updating to the latest compiler (emscripten SDK version 3.1.14), but that didn't seem to fix anything. Here's how I'm reproducing (on Linux):

(Not sure if anything from here is actually a pre-req): https://onnxruntime.ai/docs/build/web.html

```sh
 ./build.sh --cmake_generator Ninja --config Debug --skip_submodule_sync --parallel --enable_wasm_debug_info --build_wasm --update --build
node --stack-trace-limit=10000 build/Linux/Debug/onnxruntime_test_all.js --gtest_filter="SignalOpsTest.STFTFloat"
```

Reply:
Based on this I'm thinking maybe a memory allocation is failing silently and this is getting caught only in the destructor:
https://github.com/emscripten-core/emscripten/issues/11042

I'm trying to compile the test with different options to see if I can confirm.